### PR TITLE
Wikidata v0 with author description and infobox

### PR DESCRIPTION
**Pull request by RayBB**
[Wikidata Project Google Doc](https://docs.google.com/document/d/1-xAija9Pfhtwc-wCAgERHBx6GvFv40Hb-nfHd9f6SPc/edit)

In short, this PR enables us to start fetching info from wikidata and showing it on the author pages.

## How it works:
- For patrons, we only fetch missing items (and bust cache) when they visit the edit page.
  - This has the unfortunate side effect that if they add a wikidata id, it won't be loaded until the next time the edit page is visited. That is okay for this v0 because in v1 we will bulk upload the wikidata and then remove this restriction. We have this restriction now so we don't overload wikidata with requests at go live.
  - The date has also moved from under the name to the infobox for patrons.
- For librarians, we fetch missing items on the view/edit pages but only bust_cache on the edit page.
  - Cache is 30 days
  - Librarians still see the date under the name and on the infobox.

## About this first step
This PR is v0 of the wikidata project:
- (v0) 1. Way to fetch live data to handling cache_miss case
    - Also gives patrons way to reload the cached record if they have made an edit to wikidata
- (v0) 2. Render data on authors (view+edit) page
- (v0) 3. Database where we ‘cache’ wikidata entities
    - Note: Can’t use memcache since it doesn’t persist between restarts + since memcache is in memory might be too large to store there

### Technical

This PR supersedes: #8949 and #8236 - much discussion here

### Testing

Drini has tested thoroughly (see comments in this PR) but we also have nice unit tests for caching now.

### Screenshot

| Before                                                                                   | After                                                                                   |
|------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------|
|Screenshot 2024-05-02 at 00-12-27 Cory Doctorow Open Library |Screenshot 2024-05-02 at 00-12-08 Cory Doctorow Open Library |

### Stakeholders

@cdrini

### CODE REVIEW

File: openlibrary/templates/authors/infobox.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        $:render_template("covers/change", page, ".bookCover img")
+    </div>
+    <p class="short-description">
+        $if wikidata:
+            $wikidata.get_description(i18n.get_locale())
+    </p>
+    <table>
+        $if page.birth_date or page.death_date:
+            $:render_infobox_row("Born", 'birthDate', page.birth_date)
+            $:render_infobox_row("Died", 'deathDate', page.death_date)
```
Comment (cdrini) [inline openlibrary/templates/authors/infobox.html:29]: Little gotcha on the i18n here; the _ function is analyzed statically before the code is executed, so it only supports string literals.

```suggestion
            $:render_infobox_row(_("Born"), 'birthDate', page.birth_date)
            $:render_infobox_row(_("Died"), 'deathDate', page.death_date)
```

File: openlibrary/templates/authors/infobox.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        $elif page.date:
+                $:render_infobox_row("Date", '', page.date)
```
Comment (cdrini) [inline openlibrary/templates/authors/infobox.html:31]:
```suggestion
                $:render_infobox_row(_("Date"), '', page.date)
```

File: openlibrary/templates/authors/infobox.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+$if edit_view:
+    $ wikidata = page.wikidata(bust_cache=True, fetch_missing=True)
+$else:
+    $ wikidata = page.wikidata(fetch_missing=is_librarian)
+
+$def render_infobox_row(label, itemprop, value):
+    $if value:
+        <tr>
+            <td><strong>$_(label)</strong></td>
```
Comment (cdrini) [inline openlibrary/templates/authors/infobox.html:13]:
```suggestion
            <td><strong>$label</strong></td>
```

**Review by cdrini — APPROVED** [review]
Lgtm! Deploying to testing and will give a whirl there!

**Review by cdrini — COMMENTED** [review]
Tested:
- Visiting https://testing.openlibrary.org/authors/OL25788A/John_Steinbeck logged in as a librarian fetches wikidata and displays the Wikdiata short description
- Afterwards visiting https://testing.openlibrary.org/authors/OL25788A/John_Steinbeck , the short description is still visible
- Author with no Wikidata ID loads correctly: https://testing.openlibrary.org/authors/OL147566A/Klaus_von_See
- Going to an author page with a wikidata id logged out does _not_ fetch/display the wd short description: https://testing.openlibrary.org/authors/OL1121078A
- Log in as non-librarian, going to [author page](https://testing.openlibrary.org/authors/OL1123764A) does not display the wikidata description
- Clicking "Edit" does fetch/display the wd description

⚠ Modifying the short description on wikidata and reloading the author page as a non-librarian showed the new description! This suggests it re-fetched.

One bug to look into otherwise the testing went ok! Although I realized I was kind of assuming it fetched/didn't fetch the wikidata data.

**Comment by RayBB** [comment]
@cdrini I fixed it :') 
https://github.com/internetarchive/openlibrary/pull/9130/commits/fd0d267da693582717816d7fa30959d3fc6f1a10#diff-e97d8adfed291d93588f5da99576884753fdaa96ca162b04bc449948d627c18eR82

I also write a ton of tests along the way so you can feel confident that that part of the code is working.
For these few weeks until we do the bulk import and simplify it...

**Comment by tfmorris** [comment]
Is there a reason that this isn't linked to the issue #710 that it is intended to address? The side effect is that anyone who's interested in Wikidata integration and following that issue won't have been pinged to review this.

**Pull request merged**